src/main/kotlin/org/rust/lang/core/completion/RsImplTraitMemberCompletionProvider.kt

```diff
@@ -0,0 +1,127 @@
+        }
+
+        for (item in parentItems) {
+            val lookup = getCompletion(item, trait, implBlock) ?: continue
+            result.addElement(
+                lookup.withPriority(KEYWORD_PRIORITY + 1)
+            )
+        }
+    }
+}
+
+private fun getCompletion(
+    target: RsAbstractable,
+    trait: BoundElement<RsTraitItem>,
+    impl: RsImplItem
+): LookupElementBuilder? {
+    val substitution = trait.subst
+    return when (target) {
+        is RsConstant -> completeConstant(target, substitution)
+        is RsTypeAlias -> completeType(target)
+        is RsFunction -> completeFunction(target, trait, impl)
+        else -> error("unreachable")
+    }
+}
+
+private fun completeConstant(target: RsConstant, substitution: Substitution): LookupElementBuilder? {
+    val text = buildString {
+        append("const ")
+        append(target.name)
+        val type = target.typeReference?.type?.substitute(substitution) ?: return null
+        append(": ")
+        append(type.renderInsertionSafe())
+        val builder = RsDefaultValueBuilder(target.knownItems, target.containingMod, RsPsiFactory(target.project))
+        append(" = ")
+        append(builder.buildFor(type, mapOf()).text)
+        append(';')
+    }
+    return LookupElementBuilder.create(text)
+        .withIcon(RsIcons.CONSTANT)
+        .withInsertHandler { context, _ ->
+            val element = context.getElementOfType<RsConstant>() ?: return@withInsertHandler
+            val expr = element.expr ?: return@withInsertHandler
+            selectElement(expr, context.editor)
+        }
+}
+
+private fun completeType(target: RsTypeAlias): LookupElementBuilder {
+    val text = buildString {
+        append("type ")
+        append(target.name)
+        append(" = ();")
+    }
+    return LookupElementBuilder.create(text)
+        .withIcon(RsIcons.TYPE_ALIAS)
+        .withInsertHandler { context, _ ->
+            val element = context.getElementOfType<RsTypeAlias>() ?: return@withInsertHandler
+            val typeReference = element.typeReference ?: return@withInsertHandler
+            selectElement(typeReference, context.editor)
+        }
+}
+
+private fun completeFunction(
+    target: RsFunction,
+    trait: BoundElement<RsTraitItem>,
+    impl: RsImplItem
+): LookupElementBuilder {
+    val substitution = trait.subst
+    val signature = target.getSignatureText(substitution, trait, impl) ?: ""
+    val text = "$signature{\n        unimplemented!()\n    }"
+    return LookupElementBuilder
+        .create(text)
+        .withIcon(RsIcons.FUNCTION)
+        .withInsertHandler { context, _ ->
+            val element = context.getElementOfType<RsFunction>() ?: return@withInsertHandler
+            val body = element.block?.expr ?: return@withInsertHandler
+            selectElement(body, context.editor)
+        }
+        .withPresentableText("$signature { ... }")
```

Comment:
Do we actually want to use raw `$signature` as presentable text?
It may produce quite weird text
![image](https://user-images.githubusercontent.com/2539310/122600584-a8b4b480-d078-11eb-9b21-a70601749b73.png)

Reply:
Good catch. What about ignoring `where` clauses (and maybe also generics, which can also be quite long)? By doing something like this:
```kotlin
fun RsFunction.getSignatureText(
    subst: Substitution,
    trait: BoundElement<RsTraitItem>,
    impl: RsImplItem,
    withWhereClause: Boolean = true,
    withGenerics: Boolean = true
): String? {
```

Reply:
I've checked Java and Kotlin implementations
Kotlin does nothing with too long signatures, Java always skips generic "section"

Kotlin:
![image](https://user-images.githubusercontent.com/2539310/122783368-2ca9af00-d2ba-11eb-81f1-31883dfe93a2.png)
Java:
![image](https://user-images.githubusercontent.com/2539310/122783182-fff59780-d2b9-11eb-8405-6d936a166473.png)

Since Rust generics are quite long quite often, let's indeed skip generic and where parts.

Regarding suggesting `withWhereClause` and `withGenerics` - let's use a single flag (at least, for now) because currently we are not going to render only `where` clause or only generic parameters